src/libstd/process.rs

```diff
@@ -0,0 +1,835 @@
+// Copyright 2015 The Rust Project Developers. See the COPYRIGHT
+// file at the top-level directory of this distribution and at
+// http://rust-lang.org/COPYRIGHT.
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+//! Working with processes.
+
+#![unstable(feature = "process", reason = "recently added via RFC 579")]
+#![allow(non_upper_case_globals)]
+
+use core::prelude::*;
+
+use env;
+use ffi::AsOsStr;
+use fmt;
+use old_io::{self, Reader, Writer};
+use io::{Error, ErrorKind};
+use io::Result as IoResult;
+use path::AsPath;
+use libc;
+use sync::mpsc::{channel, Receiver};
+use sys::fs::FileDesc;
+use sys::process2::Process as ProcessImp;
+use sys::process2::Command as CommandImp;
+use sys;
+use sys_common::{AsInner, AsInnerMut};
+use thread::Thread;
+use vec::Vec;
+
+/// Signal a process to exit immediately, forcibly killing it. Corresponds to
+/// SIGKILL on unix platforms.
+#[cfg(windows)] const MustDieSignal: int = 9;
+/// Signal a process to exit immediately, forcibly killing it. Corresponds to
+/// SIGKILL on unix platforms.
+#[cfg(not(windows))] const MustDieSignal: int = libc::SIGKILL as int;
+
+/// Representation of a running or exited child process.
+///
+/// This structure is used to represent and manage child processes. A child
+/// process is created via the `Command` struct, which configures the spawning
+/// process and can itself be constructed using a builder-style interface.
+///
+/// # Example
+///
+/// ```should_fail
+/// use std::old_io::Command;
+///
+/// let mut child = match Command::new("/bin/cat").arg("file.txt").spawn() {
+///     Ok(child) => child,
+///     Err(e) => panic!("failed to execute child: {}", e),
+/// };
+///
+/// let contents = child.stdout.as_mut().unwrap().read_to_end();
+/// assert!(child.wait().unwrap().success());
+/// ```
+pub struct Child {
+    handle: ProcessImp,
+
+    /// None until wait() is called.
+    status: Option<ExitStatus>,
+
+    stdin: Option<ChildStdin>,
+    stdout: Option<ChildStdout>,
+    stderr: Option<ChildStderr>,
+}
+
+/// A handle to a child procesess's stdin
+pub struct ChildStdin {
+    inner: FileDesc
```

Comment:
Now that #21936 has landed we've also got [`unix::fd::FileDesc`](https://github.com/rust-lang/rust/blob/master/src/libstd/sys/unix/fd.rs) and [`windows::handle::Handle`](https://github.com/rust-lang/rust/blob/master/src/libstd/sys/windows/handle.rs) for representing the raw representation. Right now we use a bunch of `osfhandle` stuff which we shouldn't be using (because we're using file descriptors). I'm not sure of the precise API to read/write from the handles we have for the child on Windows, but this may be able to contain just a wrapper around `sys::fs::File` (which has a Handle/FileDesc appropriately).

In general though this is a pretty minor implementation detail we can tweak at any time (so long as the API is right, which it will be), so it's not too major.